src/Microsoft.FeatureManagement/ContextualFeatureVariantAssignerEvaluator.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.FeatureManagement
+{
+    /// <summary>
+    /// Provides a performance efficient method of evaluating IContextualFeatureVariantAssigner&lt;T&gt; without knowing what the generic type parameter is.
+    /// </summary>
+    sealed class ContextualFeatureVariantAssignerEvaluator : IContextualFeatureVariantAssigner<object>
+    {
+        private IFeatureVariantAssignerMetadata _filter;
+        private Func<object, FeatureVariantAssignmentContext, object, CancellationToken, ValueTask<FeatureVariant>> _evaluateFunc;
+
+        public ContextualFeatureVariantAssignerEvaluator(IFeatureVariantAssignerMetadata assigner, Type appContextType)
+        {
+            if (assigner == null)
+            {
+                throw new ArgumentNullException(nameof(assigner));
+            }
+
+            if (appContextType == null)
+            {
+                throw new ArgumentNullException(nameof(appContextType));
+            }
+
+            Type targetInterface = GetContextualAssignerInterface(assigner, appContextType);
+
+            //
+            // Extract IContextualFeatureFilter<T>.EvaluateAsync method.
+            if (targetInterface != null)
+            {
+                MethodInfo evaluateMethod = targetInterface.GetMethod(nameof(IContextualFeatureVariantAssigner<object>.AssignVariantAsync), BindingFlags.Public | BindingFlags.Instance);
+
+                _evaluateFunc = TypeAgnosticEvaluate(assigner.GetType(), evaluateMethod);
+            }
+
+            _filter = assigner;
+        }
+
+        public ValueTask<FeatureVariant> AssignVariantAsync(FeatureVariantAssignmentContext assignmentContext, object context, CancellationToken cancellationToken)
+        {
+            if (_evaluateFunc == null)
+            {
+                return new ValueTask<FeatureVariant>((FeatureVariant)null);
+            }
+
+            return _evaluateFunc(_filter, assignmentContext, context, cancellationToken);
+        }
+
+        public static bool IsContextualFilter(IFeatureVariantAssignerMetadata assigner, Type appContextType)
+        {
+            return GetContextualAssignerInterface(assigner, appContextType) != null;
```

Comment:
>assigner, appContextType [](http://example.com/codeflow?start=50&length=24)

Do we need null checks for these arguments?

Reply:
Yes, thanks.